Supervise cron process for better logging/signal handling/environment management

ndarilek opened the pull request.

davissp14 commented on internal/flybarman/node.go:
```diff
@@ -140,7 +140,7 @@ wal_retention_policy = main
 
 	if _, err := os.Stat(n.BarmanCronFile); os.IsNotExist(err) {
-		barmanCronFileContent := `* * * * * /usr/bin/barman cron
+		barmanCronFileContent := `* * * * * . $HOME/.profile; /usr/bin/barman cron >> /proc/1/fd/1 2>/proc/1/fd/2
```
Would have to test this to confirm, but now that cron is wrapped within a supervisor process we might not actually need to pipe the logs to the main process.

ndarilek replied:
I just spun up a debian:12 docker image, installed cron, added the following crontab entry, and ran `cron -f`:
```
* * * * * root echo test
```

Nothing appeared on stdout, so I think the redirects are still needed.

davissp14 replied:
Aight, so this should do the trick.  Basically, you could add this to the bin directory in this project and then modify the file the cronjob calls to `/usr/bin/path-to-barman-cron-file.sh`.  

This will identify the cron supervisor process id, run the `barman cron` command and then log the result. 

```
#!/bin/bash

# Identify the process id of the cron supervisor process
cron_pid=$(pgrep -f '/usr/sbin/cron')
if [ -z "$cron_pid" ]; then
    # Send the error message to main process's stderr
    echo "Failed to resolve cron process id" >> /proc/1/fd/2
    exit 1
fi

barman_output=$(barman cron 2>&1)
barman_exit_code=$?

# Log the result of barman cron
if [ $barman_exit_code -ne 0 ]; then
  echo "Barman cron failed with exit code $barman_exit_code: $barman_output" >> /proc/$cron_pid/fd/2
  exit 1
else
  echo "Barman cron succeeded: $barman_output" >> /proc/$cron_pid/fd/1
fi
```

ndarilek replied:
Thanks for doing this legwork. My launchpad issues cleared up, so I was finally able to build this and put it through its paces as well as I could.

I've pushed what's on this branch to [nolan/postgres-flex](https://hub.docker.com/r/nolan/postgres-flex) to make testing a bit easier. Unfortunately we don't seem to support `fly pg barman create --image-ref ...` for easier testing, unless I'm missing something. If no one else gets to that, I can do it on Monday.

davissp14 replied:
Yeah, this is annoying.  You have to manually update the machine:

```
fly machines <machine-id> --image <image-ref>
```

ndarilek commented:
I'd appreciate it if someone could test this. There was an incident with Launchpad this morning that blocked me from downloading repmgr and building the image. It seems to be resolved but I still can't get the file via curl. Going to keep trying in the hope that it's a caching issue somewhere and gets resolved soon, I'll remove this paragraph if I manage to test this today.

In particular, I'm curious if cron even logs the output of its jobs to stdout/stderr or if it sends it to syslog/the journal instead. How hard is it to trigger a barman failure of some sort to check that?

ndarilek commented:
Ah, NM, we're redirecting the job output itself to stdout/stderr so it doesn't matter what cron itself does.

davissp14 commented:
@ndarilek I will take a look!

ndarilek commented:
I updated the machine, removed the original crontab so it regenerated, and confirmed that the script appeared to run successfully (I.e. just ran silently without error.)

I then updated the barman command to `echo 2>& Test`, manually set the exit code to 1, and got the following in the machine logs:

```
2024-06-25T16:57:01Z app[d8dd957fedde38] dfw [info]cron   | Barman cron failed with exit code 1: Test
```

So I think this one's ready. I did just push another commit to always rewrite the crontab--I can't imagine we'll change it often but if we do then I doubt having the slight performance boost of checking existence first would be a game-changer. Let me know if I'm missing something WRT that.

ndarilek commented:
Also updated the nolan/postgres-flex image with the most recent changes (I.e. not checking for crontab existence.)
